File: android/app/src/main/java/app/networked/ai/IconSwitcherPlugin.java
package app.networked.ai;

import android.content.ComponentName;
import android.content.pm.PackageManager;

import com.getcapacitor.Plugin;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "IconSwitcher")
public class IconSwitcherPlugin extends Plugin {

    private final String[] aliases = {
        "icon_1",
        "icon_2",
        "icon_3",
        "icon_4",
        "icon_5",
        "icon_6",
        "icon_7",
        "icon_8",
        "icon_9",
        "icon_10",
        "icon_11",
        "icon_12",
        "icon_13",
        "icon_14",
        "icon_15",
        "icon_16"
    };

    @PluginMethod
    public void setIcon(PluginCall call) {
        String iconName = call.getString("iconName");

        if (iconName == null || iconName.isEmpty()) {
            call.reject("iconName is required");
            return;
        }

        String pkg = getContext().getPackageName();
        PackageManager pm = getContext().getPackageManager();

        // disable all icons
        for (String alias : aliases) {
            ComponentName component = new ComponentName(pkg, pkg + "." + alias);
            pm.setComponentEnabledSetting(
                component,
                PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP
            );
        }

        // enable selected icon
        ComponentName selected = new ComponentName(pkg, pkg + "." + iconName);
        pm.setComponentEnabledSetting(
            selected,
            PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
            PackageManager.DONT_KILL_APP
        );

        call.resolve();
    }

    @PluginMethod
    public void getAvailableIcons(PluginCall call) {
        try {
            JSArray array = new JSArray();

            for (String alias : aliases) {
                array.put(alias);
            }

            JSObject result = new JSObject();
            result.put("icons", array);

            call.resolve(result);

        } catch (Exception e) {
            call.reject("Error getting icons");
        }
    }
}

File: ios/App/App/IconSwitcherPlugin.swift
//  Created by Ravi Gaud on 25/02/26.

import UIKit
import Capacitor

@objc(IconSwitcherPlugin)
public class IconSwitcherPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "IconSwitcherPlugin"
    public let jsName = "IconSwitcher"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setIcon", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAvailableIcons", returnType: CAPPluginReturnPromise)
    ]

    private let aliases: [String] = [
        "icon_1",
        "icon_2",
        "icon_3",
        "icon_4",
        "icon_5",
        "icon_6",
        "icon_7",
        "icon_8",
        "icon_9",
        "icon_10",
        "icon_11",
        "icon_12",
        "icon_13",
        "icon_14",
        "icon_15",
        "icon_16"
    ]

    @objc func setIcon(_ call: CAPPluginCall) {
        guard let iconName = call.getString("iconName") else {
            call.reject("iconName is required")
            return
        }

        DispatchQueue.main.async {

            guard UIApplication.shared.supportsAlternateIcons else {
                call.reject("Alternate icons not supported on this device")
                return
            }

            // Allow reset to primary icon if empty string passed
            let targetIcon = iconName.isEmpty ? nil : iconName

            // Optional: Validate icon exists (if not resetting)
            if let name = targetIcon, !self.aliases.contains(name) {
                call.reject("Invalid icon name")
                return
            }

            UIApplication.shared.setAlternateIconName(targetIcon) { error in
                if let error = error {
                    call.reject("Failed to change icon: \(error.localizedDescription)")
                } else {
                    call.resolve()
                }
            }
        }
    }

    @objc func getAvailableIcons(_ call: CAPPluginCall) {
        call.resolve([ "icons": aliases ])
    }
}
